sql: CockroachDB panics when executing an ill-form AS OF SYSTEM TIME expression

**Describe the problem**

The latest version of the CockroachDB v24.2.0 crashes when executing the following query:

```sql
START TRANSACTION AS OF SYSTEM TIME ('abc' IS OF ( ident ) );
```

**To Reproduce**

1. In operating system Ubuntu 20.04 LTS, download the pre-build CockroachDB binaries (v24.2.0) from [link](https://www.cockroachlabs.com/docs/releases/#v24-2)
2. Run ./cockroach demo, and then paste the PoC query to the cockroach cli environment.
3. Observe the crash and log the stack information.

**Expected behavior**
The CockroachDB should return error from the statement. 

**Additional data / screenshots**

This is the stack frame from the crash (v24.2.0):

```
*
* ERROR: a SQL panic has occurred while executing the following statement:
* START TRANSACTION AS OF SYSTEM TIME ('abc' IS OF (ident))
*
*
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:961
*   | [...repeated from below...]
* Wraps: (2) while executing: START TRANSACTION AS OF SYSTEM TIME ('_' IS OF (_))
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:961
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:261
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:881
*   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).ReadTimestamp
*   | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:401
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).getByName
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/leased_descriptors.go:105
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getNonVirtualDescriptorID.func6
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:580
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getNonVirtualDescriptorID
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:620
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.getDescriptorByName
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:398
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.ByNameGetter.Database
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/getters.go:249
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.getObjectPrefix
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/helpers.go:97
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.PrefixAndType
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/helpers.go:110
*   | github.com/cockroachdb/cockroach/pkg/sql.(*schemaResolver).LookupObject
*   | 	github.com/cockroachdb/cockroach/pkg/sql/schema_resolver.go:176
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExisting
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:380
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExistingObject
*   | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:184
*   | github.com/cockroachdb/cockroach/pkg/sql.(*schemaResolver).ResolveType
*   | 	github.com/cockroachdb/cockroach/pkg/sql/schema_resolver.go:333
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.ResolveType
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_name.go:140
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*IsOfTypeExpr).TypeCheck
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:1552
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ParenExpr).TypeCheck
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:1652
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/asof.Eval
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/asof/as_of.go:187
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).EvalAsOfTimestamp
*   | 	github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2083
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).beginTransactionTimestampsAndReadMode
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2459
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInNoTxnState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2539
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:138
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2316
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2321
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2238
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:963
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:253
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:1036
*   | runtime.goexit
*   | 	src/runtime/asm_arm64.s:1222
* Wraps: (4) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.errorString
*
*
* ERROR: Queued as error 136b13de825046fcae7dc18fe9cb04e3
*
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1693f38]

goroutine 5098 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0x40062b0488, {0x7b90c90, 0x4001b48fc0}, {0x57b7b80?, 0xb6b5630?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1234 +0x160
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x7b90c90, 0x4001b48fc0}, {0x4002baec40?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:961 +0x4c
panic({0x57b7b80?, 0xb6b5630?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).ReadTimestamp(0x0?)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:401 +0x28
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).getByName(0x40023b50a8, {0x7b90c90, 0x4001b48fc0}, {0x7b5b560, 0x0}, 0x0, 0x0, {0x400109c039, 0x4})
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/leased_descriptors.go:105 +0x168
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getNonVirtualDescriptorID.func6()
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:580 +0x138
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getNonVirtualDescriptorID(0x0?, {0x7b90c90?, 0x4001b48fc0?}, 0x400366e600?, {0x0?, 0x0?}, {0x0?, 0x0?}, {0x400109c039?, 0x17f496740a088d03?}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:620 +0x37c
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.getDescriptorByName({0x7b90c90, 0x4001b48fc0}, 0x0, 0x40023b5080, {0x0, 0x0}, {0x0, 0x0}, {0x400109c039, 0x4}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:398 +0x128
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.ByNameGetter.Database({{0x0?, 0x40023b5080?}, {{0x0?, 0x0?}, {0x0?, 0x0?, 0x0?, 0x0?}, {0x3?, 0x8d?, ...}}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/getters.go:249 +0x9c
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.getObjectPrefix({0x7b90c90, 0x4001b48fc0}, {{0x0?, 0x40023b5080?}, {{0x0?, 0x0?}, {0x0?, 0x0?, 0x0?, 0x0?}, ...}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/helpers.go:97 +0xc8
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.PrefixAndType({0x7b90c90, 0x4001b48fc0}, {{0x0?, 0x40023b5080?}, {{0x32?, 0x0?}, {0x0?, 0x0?, 0x0?, 0x0?}, ...}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/helpers.go:110 +0xe4
github.com/cockroachdb/cockroach/pkg/sql.(*schemaResolver).LookupObject(0x40062b0b38, {0x7b90c90, 0x4001b48fc0}, {0x1, 0x0, 0x0, 0x0, 0x0, 0x2, 0x0}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/schema_resolver.go:176 +0x648
github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExisting({0x7b90c90, 0x4001b48fc0}, 0x4005fed2c0, {0xffff42512b08, 0x40062b0b38}, {0x1, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:380 +0x4cc
github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExistingObject({0x7b90c90, 0x4001b48fc0}, {0x7bf9b90, 0x40062b0b38}, 0x4005fed2c0, {0x1, 0x0, 0x0, 0x0, 0x0, ...})
	github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:184 +0x15c
github.com/cockroachdb/cockroach/pkg/sql.(*schemaResolver).ResolveType(0x40062b0b38, {0x7b90c90, 0x4001b48fc0}, 0x4005fed2c0)
	github.com/cockroachdb/cockroach/pkg/sql/schema_resolver.go:333 +0x80
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.ResolveType({0x7b90c90?, 0x4001b48fc0?}, {0x7b3dc00?, 0x4005fed2c0?}, {0x7b5b2e0?, 0x40062b0b38?})
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_name.go:140 +0x80
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*IsOfTypeExpr).TypeCheck(0x400832c460, {0x7b90c90, 0x4001b48fc0}, 0x40062b0f50, 0x4008c766f8?)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:1552 +0x140
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ParenExpr).TypeCheck(0x400871c3a8, {0x7b90c90?, 0x4001b48fc0?}, 0x34372d8?, 0x4008c76738?)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:1652 +0x38
github.com/cockroachdb/cockroach/pkg/sql/sem/asof.Eval({0x7b90c90, 0x4001b48fc0}, {{0x7b91438?, 0x400871c3a8?}}, 0x40062b0f50, 0x40062b1040, {0x0, 0x0, 0x30?})
	github.com/cockroachdb/cockroach/pkg/sql/sem/asof/as_of.go:187 +0x4f8
github.com/cockroachdb/cockroach/pkg/sql.(*planner).EvalAsOfTimestamp(0x40062b0488?, {0x7b90c90?, 0x4001b48fc0?}, {{0x7b91438?, 0x400871c3a8?}}, {0x0?, 0x0?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2083 +0x70
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).beginTransactionTimestampsAndReadMode(0x40062b0488, {0x7b90c90, 0x4001b48fc0}, 0x4005fed300)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2459 +0x234
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInNoTxnState(0x40062b0488, {0x7b90c90, 0x4001b48fc0}, {{0x7bb7f40, 0x4005fed300}, {0x0, 0x0, 0x0}, {0x400634d309, 0x3c}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2539 +0x228
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x40062b0488, {0x7b90c90, 0x4001b48fc0}, {{0x7bb7f40, 0x4005fed300}, {0x0, 0x0, 0x0}, {0x400634d309, 0x3c}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:138 +0x388
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x4008c777f8, 0x4008c778a8, 0x40062b0488, 0x4008c774d0, 0x4008c77930, 0x4008c77920, 0x4008c77950, 0x4008c77968)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2316 +0x360
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x40062b0488)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2321 +0x764
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x40062b0488, {0x7b90c90, 0x4001b48fc0}, 0x4001b91c80?, 0x4009e2dc98?, 0x40061b6530?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2238 +0x168
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0x4002c04280?, {0x7b90c90?, 0x4001b48fc0?}, {0x400a476ea0?}, 0x2?, 0x4001b48d80?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:963 +0xb4
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0x40026f4008, {0x7b90c90, 0x4001b48fc0}, {0x0, 0x4, {0x1, {0x62212c7, 0x3}, {0x4003a7de80, 0xf}, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:253 +0x344
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4()
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:1036 +0xcc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl in goroutine 5186
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:1033 +0x498
```

**Environment:**

- CockroachDB version: v24.2.0
- Server OS: Ubuntu 20.04 LTS
- Client app: CockroachDB demo command. (./cockroach demo)
- Detailed CockroachDB version output:

```
cockroach version details:
Build Tag:        v24.2.0
Build Time:       2024/08/08 01:50:30
Distribution:     CCL
Platform:         linux arm64 (aarch64-unknown-linux-gnu)
Go Version:       go1.22.5 X:nocoverageredesign
C Compiler:       gcc 6.5.0
Build Commit ID:  953d09cce5a01cc0053e138516d80a27454f027c
Build Type:       release
```



Jira issue: CRDB-42137

Epic CRDB-40419

--- a/pkg/sql/logictest/testdata/logic_test/txn_as_of
+++ b/pkg/sql/logictest/testdata/logic_test/txn_as_of
@@ -280,3 +280,15 @@ SET TRANSACTION AS OF SYSTEM TIME '0'
 
 statement ok
 ROLLBACK
+
+statement error AS OF SYSTEM TIME: only constant expressions or follower_read_timestamp are allowed
+BEGIN TRANSACTION AS OF SYSTEM TIME ('abc' IS OF (ident))
+
+statement ok
+BEGIN
+
+statement error AS OF SYSTEM TIME: only constant expressions or follower_read_timestamp are allowed
+SET TRANSACTION AS OF SYSTEM TIME ('abc' IS OF (ident))
+
+statement ok
+ROLLBACK

--- a/pkg/sql/schema_resolver.go
+++ b/pkg/sql/schema_resolver.go
@@ -122,7 +122,6 @@ func (sr *schemaResolver) byNameGetterBuilder() descs.ByNameGetterBuilder {
 func (sr *schemaResolver) LookupObject(
 	ctx context.Context, flags tree.ObjectLookupFlags, dbName, scName, obName string,
 ) (found bool, prefix catalog.ResolvedObjectPrefix, desc catalog.Descriptor, err error) {
-
 	// Check if we are looking up a type which matches a built-in type in
 	// CockroachDB but is an extension type on the public schema in PostgreSQL.
 	if flags.DesiredObjectKind == tree.TypeObject && scName == catconstants.PublicSchemaName {

--- a/pkg/sql/sem/asof/BUILD.bazel
+++ b/pkg/sql/sem/asof/BUILD.bazel
@@ -19,6 +19,7 @@ go_library(
         "//pkg/util/hlc",
         "@com_github_cockroachdb_apd_v3//:apd",
         "@com_github_cockroachdb_errors//:errors",
+        "@com_github_lib_pq//oid",
     ],
 )
 

--- a/pkg/sql/sem/asof/as_of.go
+++ b/pkg/sql/sem/asof/as_of.go
@@ -20,6 +20,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/duration"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/errors"
+	"github.com/lib/pq/oid"
 )
 
 // FollowerReadTimestampFunctionName is the name of the function which can be
@@ -132,6 +133,14 @@ func Eval(
 	defer scalarProps.Restore(*scalarProps)
 	scalarProps.Require("AS OF SYSTEM TIME", tree.RejectSpecial|tree.RejectSubqueries)
 
+	// Disable type resolution. Since type resolution requires a transaction, but
+	// this expression is being evaluated before a transaction begins, resolving
+	// any type should result in an error. There is no valid AS OF SYSTEM TIME
+	// expression that requires type resolution.
+	origTypeResolver := semaCtx.GetTypeResolver()
+	semaCtx.TypeResolver = &asOfTypeResolver{errFactory: newInvalidExprError}
+	defer func() { semaCtx.TypeResolver = origTypeResolver }()
+
 	var ret eval.AsOfSystemTime
 
 	// In order to support the follower reads feature we permit this expression
@@ -202,6 +211,29 @@ func Eval(
 		return eval.AsOfSystemTime{}, errors.Wrap(err, "AS OF SYSTEM TIME")
 	}
 	return ret, nil
+}
+
+// asOfTypeResolver is a type resolver that always returns an error. It is used
+// to block type resolution while evaluating the AS OF SYSTEM TIME expression.
+type asOfTypeResolver struct {
+	// errFactory is a function that returns the error to be returned by the
+	// type resolver. Using a closure lets us avoid instantiating the error
+	// unless something actually tries to resolve a type.
+	errFactory func() error
+}
+
+var _ tree.TypeReferenceResolver = (*asOfTypeResolver)(nil)
+
+// ResolveType implements the tree.TypeReferenceResolver interface.
+func (r *asOfTypeResolver) ResolveType(
+	ctx context.Context, name *tree.UnresolvedObjectName,
+) (*types.T, error) {
+	return nil, r.errFactory()
+}
+
+// ResolveTypeByOID implements the tree.TypeReferenceResolver interface.
+func (r *asOfTypeResolver) ResolveTypeByOID(ctx context.Context, oid oid.Oid) (*types.T, error) {
+	return nil, r.errFactory()
 }
 
 // DatumToHLCUsage specifies which statement DatumToHLC() is used for.

